crates/noirc_evaluator/src/ssa/node.rs

```diff
@@ -350,642 +357,800 @@ impl Instruction {
+    where
+        F: FnMut(&SsaContext, NodeId) -> NodeEval,
+    {
+        match &self.operator {
+            Operation::Binary(binary) => {
+                return binary.evaluate(ctx, self.id, self.res_type, eval_fn)
+            }
+            Operation::Cast(value) => {
+                if let Some(l_const) = ctx.get_const_value(*value) {
+                    if self.res_type == ObjectType::NativeField {
+                        return NodeEval::Const(FieldElement::from(l_const), self.res_type);
+                    }
+                    return NodeEval::Const(
+                        FieldElement::from(l_const % (1_u128 << self.res_type.bits())),
+                        self.res_type,
+                    );
+                }
+            }
+            Operation::Not(value) => {
+                if let Some(l_const) = ctx.get_const_value(*value) {
+                    return NodeEval::Const(FieldElement::from(!l_const), self.res_type);
+                }
+            }
+            Operation::Phi { .. } => (), //Phi are simplified by simply_phi() later on; they must not be simplified here
+            _ => (),
         }
-        if self.lhs != NodeId::dummy() {
-            return vec![self.lhs];
+        NodeEval::VarOrInstruction(self.id)
+    }
+
+    // Simplifies trivial Phi instructions by returning:
+    // None, if the instruction is unreachable or in the root block and can be safely deleted
+    // Some(id), if the instruction can be replaced by the node id
+    // Some(ins_id), if the instruction is not trivial
+    pub fn simplify_phi(ins_id: NodeId, phi_arguments: &[(NodeId, BlockId)]) -> Option<NodeId> {
+        let mut same = None;
+        for op in phi_arguments {
+            if Some(op.0) == same || op.0 == ins_id {
+                continue;
+            }
+            if same.is_some() {
+                //no simplification
+                return Some(ins_id);
+            }
+
+            same = Some(op.0);
         }
-        if self.rhs != NodeId::dummy() {
-            return vec![self.rhs];
+        //if same.is_none()  => unreachable phi or in root block, can be replaced by ins.lhs (i.e the root) then.
+        same
+    }
+
+    /// Delete this instruciton by mutating it into a Operation::Nop
+    pub fn delete(&mut self) {
+        self.operator = Operation::Nop;
+    }
+
+    pub fn is_deleted(&self) -> bool {
+        self.operator == Operation::Nop || self.replacement.is_some()
+    }
+
+    pub fn standard_form(&mut self) {
+        if let Operation::Binary(binary) = &mut self.operator {
+            if let BinaryOp::Constrain(op) = &binary.operator {
+                match op {
+                    ConstrainOp::Eq => {
+                        if binary.lhs == binary.rhs {
+                            self.delete();
+                            return;
+                        }
+                    }
+                    // TODO: Why are we asserting here?
```

Comment:
A neq should be perfectly valid with identical operands, it should just fail to verify or we should issue a warning for it

Reply:
If they're not equal, we should fail verification rather than crashing the compiler

Reply:
I agree we should not crash the compiler, so we should return a compilation error saying that the constrain could not be satisfied.
What we should not do is to output an invalid circuit, so we must return an error..

Reply:
No this is incorrect; the circuit is not valid because there is no valid witness that can satisfy the constraints.
Noir should not generate a proof that fails to verify, because that means it is an invalid proof.
We need people to trust Noir and this means at least it outputs proper proofs. So relying on a proof that fail to verify is not a good idea.

Reply:
Are you suggesting we only output proofs that successfully verify? Unless I'm missing something this seems impossible in the general case. 

I agree we should provide better errors/warnings where possible, which is why I've opened issue #231. I disagree that crashing the compiler with an assert is better for users than failing verification. Ideally we'd have full error reporting with source locations, etc. This is large enough that it should be its own discussion/PR for that issue.

Reply:
Yes we should definitively only output proofs that verify properly, and I don't see why this would not be possible.

I see that we disagree on the assert vs verification, I agree on the ideal case and also agree that this is large enough to discuss this outside.
So I suggest to put back the assert so that the refactoring does not change the behavior on this point.
Note that the SSA pass is not enabled so the asserts do not crash the compiler. All we will need to do is to agree on this point (and may be others as well :) ) and implement what we agree on **before** we enable the SSA pass.

Reply:
How would we always only output a proof that successfully verifies for `constrain x == y` for some private x and y? 
I'd like to keep the current interpreter's behavior so that `constrain x == x` will fail verification instead of crashing when run. A proper error message would be better than either, but I consider an assert fail/panic as something that a bug report should be expected for rather than a normal result of compilation.

Reply:
> How would we always only output a proof that successfully verifies for `constrain x == y` for some private x and y?

This is easy, we generate an error if `x !=y`.
May be I was not clear so let me clarify this: ouput a proof that fails verification is a NO-GO for me.
I know this can currently happen but it should be fixed when we have the time rather than introducing more cases for bad proofs.

Reply:
Comment removed